client/extensions/woocommerce/state/data-layer/products/index.js

```diff
@@ -2,38 +2,58 @@
  * Internal dependencies
  */
 import { dispatchWithProps } from 'woocommerce/state/helpers';
-import { post } from 'woocommerce/state/data-layer/request/actions';
+import { get, post } from 'woocommerce/state/data-layer/request/actions';
 import { setError } from 'woocommerce/state/sites/status/wc-api/actions';
 import { productUpdated } from 'woocommerce/state/sites/products/actions';
 import {
 	WOOCOMMERCE_PRODUCT_CREATE,
+	WOOCOMMERCE_PRODUCT_REQUEST,
 } from 'woocommerce/state/action-types';
 
-export function handleProductCreate( store, action ) {
+export default {
+	[ WOOCOMMERCE_PRODUCT_CREATE ]: [ handleProductCreate ],
+	[ WOOCOMMERCE_PRODUCT_REQUEST ]: [ handleProductRequest ],
+};
+
+/**
+ * Wraps an action with a product update action.
+ *
+ * @param {Number} siteId The id of the site upon which the request should be made.
+ * @param {Object} originatingAction The action which precipitated this request.
+ * @param {Object} successAction Will be dispatched with extra props: { sentData, receivedData }
+ * @param {Object} [sentData] The sentData to be embedded in the successAction along with receivedData.
+ * @return {Function} Curried function to be dispatched.
+ */
+function updatedAction( siteId, originatingAction, successAction, sentData ) {
+	return ( dispatch, getState, { data: receivedData } ) => {
+		dispatch( productUpdated( siteId, receivedData, originatingAction ) );
+
+		const props = { sentData, receivedData };
+		dispatchWithProps( dispatch, getState, successAction, props );
+	};
+}
+
+export function handleProductCreate( { dispatch }, action ) {
 	const { siteId, product, successAction, failureAction } = action;
 
 	// Filter out any id we might have.
 	const { id, ...productData } = product;
 
 	if ( typeof id === 'number' ) {
-		store.dispatch( setError( siteId, action, {
+		dispatch( setError( siteId, action, {
 			message: 'Attempting to create a product which already has a valid id.',
 			product,
 		} ) );
 		return;
 	}
 
-	const updatedAction = ( dispatch, getState, { data } ) => {
-		dispatch( productUpdated( siteId, data, action ) );
-
-		const props = { sentData: action.product, receivedData: data };
-		dispatchWithProps( dispatch, getState, successAction, props );
-	};
-
-	store.dispatch( post( siteId, 'products', productData, updatedAction, failureAction ) );
+	const updatedSuccessAction = updatedAction( siteId, action, successAction, action.product );
```

Comment:
`product` is defined on [line 37](https://github.com/Automattic/wp-calypso/pull/15586/files#diff-c4ffe9714e2ffbf73ac53f86c1e8d04bR37) 😄